# Insert Libraries

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns

# Import Dataset

In [2]:
dataset=pd.read_csv("C://Users//mehta//Desktop//MENTAL HEALTH DATASET.csv")

In [3]:
dataset.head()

Timestamp  Age  Gender         Country state self_employed  \
0  27-08-2014 11:29   37  Female   United States    IL           NaN   
1  27-08-2014 11:29   44       M   United States    IN           NaN   
2  27-08-2014 11:29   32    Male          Canada   NaN           NaN   
3  27-08-2014 11:29   31    Male  United Kingdom   NaN           NaN   
4  27-08-2014 11:30   31    Male   United States    TX           NaN   

  family_history treatment work_interfere    no_employees  ...  \
0             No       Yes          Often          Jun-25  ...   
1             No        No         Rarely  More than 1000  ...   
2             No        No         Rarely          Jun-25  ...   
3            Yes       Yes          Often          26-100  ...   
4             No        No          Never         100-500  ...   

                leave mental_health_consequence phys_health_consequence  \
0       Somewhat easy                        No                      No   
1          Don't know                     Maybe                      No   
2  Somewhat difficult                        No                      No   
3  Somewhat difficult                       Yes                     Yes   
4          Don't know                        No                      No   

      coworkers supervisor mental_health_interview phys_health_interview  \
0  Some of them        Yes                      No                 Maybe   
1            No         No                      No                    No   
2           Yes        Yes                     Yes                   Yes   
3  Some of them         No                   Maybe                 Maybe   
4  Some of them        Yes                     Yes                   Yes   

  mental_vs_physical obs_consequence comments  
0                Yes              No      NaN  
1         Don't know              No      NaN  
2                 No              No      NaN  
3                 No             Yes      NaN  
4         Don't know              No      NaN  

[5 rows x 27 columns]

# Data Cleaning

Print all the name of the columns present in the dataset.

In [4]:
dataset.columns

Index(['Timestamp', 'Age', 'Gender', 'Country', 'state', 'self_employed',
       'family_history', 'treatment', 'work_interfere', 'no_employees',
       'remote_work', 'tech_company', 'benefits', 'care_options',
       'wellness_program', 'seek_help', 'anonymity', 'leave',
       'mental_health_consequence', 'phys_health_consequence', 'coworkers',
       'supervisor', 'mental_health_interview', 'phys_health_interview',
       'mental_vs_physical', 'obs_consequence', 'comments'],
      dtype='object')

Delete the columns which are not important.

In [5]:
dataset.drop(["benefits","anonymity","leave","coworkers","supervisor"], inplace = True, axis=1)

In [6]:
dataset.shape

(1259, 22)

There are 1259 rows and 22 columns.

In [7]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1259 entries, 0 to 1258
Data columns (total 22 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Timestamp                  1259 non-null   object
 1   Age                        1259 non-null   int64 
 2   Gender                     1259 non-null   object
 3   Country                    1259 non-null   object
 4   state                      744 non-null    object
 5   self_employed              1241 non-null   object
 6   family_history             1259 non-null   object
 7   treatment                  1259 non-null   object
 8   work_interfere             995 non-null    object
 9   no_employees               1259 non-null   object
 10  remote_work                1259 non-null   object
 11  tech_company               1259 non-null   object
 12  care_options               1259 non-null   object
 13  wellness_program           1259 non-null   object
 14  seek_hel

Check if there are any null values

In [8]:
dataset.isnull().sum()

Timestamp                       0
Age                             0
Gender                          0
Country                         0
state                         515
self_employed                  18
family_history                  0
treatment                       0
work_interfere                264
no_employees                    0
remote_work                     0
tech_company                    0
care_options                    0
wellness_program                0
seek_help                       0
mental_health_consequence       0
phys_health_consequence         0
mental_health_interview         0
phys_health_interview           0
mental_vs_physical              0
obs_consequence                 0
comments                     1095
dtype: int64

So, there are 515 null values for column state, 264 for work_inference,18 in self-employed and 1095 in comments column. All the others are not null.

In [37]:
dataset = dataset.fillna({
    'work_interfere': 'No work',
    'comments': 'No comments',
    'self_employed':'No',
    'state':'0'
})


In [38]:
dataset.isnull().sum()

Timestamp                    0
Age                          0
Gender                       0
Country                      0
state                        0
self_employed                0
family_history               0
treatment                    0
work_interfere               0
no_employees                 0
remote_work                  0
tech_company                 0
care_options                 0
wellness_program             0
seek_help                    0
mental_health_consequence    0
phys_health_consequence      0
mental_health_interview      0
phys_health_interview        0
mental_vs_physical           0
obs_consequence              0
comments                     0
dtype: int64

Now, there are no null values in the dataset

Now check every column and the valaues present in them. If there are any outliers, then remove them

# Age 

In [11]:
Q1 = dataset.quantile(0.25)
Q3= dataset.quantile(0.75)
IQR = Q3 - Q1
print(IQR)

Age    9.0
dtype: float64


In [12]:
print(dataset<(Q1-1.5*IQR))or(dataset>(Q3+1.5*IQR))

        Age  Country  Gender  Timestamp  care_options  comments  \
0     False    False   False      False         False     False   
1     False    False   False      False         False     False   
2     False    False   False      False         False     False   
3     False    False   False      False         False     False   
4     False    False   False      False         False     False   
...     ...      ...     ...        ...           ...       ...   
1254  False    False   False      False         False     False   
1255  False    False   False      False         False     False   
1256  False    False   False      False         False     False   
1257  False    False   False      False         False     False   
1258  False    False   False      False         False     False   

      family_history  mental_health_consequence  mental_health_interview  \
0              False                      False                    False   
1              False                      F

Age  Country  Gender  Timestamp  care_options  comments  \
0     False    False   False      False         False     False   
1     False    False   False      False         False     False   
2     False    False   False      False         False     False   
3     False    False   False      False         False     False   
4     False    False   False      False         False     False   
...     ...      ...     ...        ...           ...       ...   
1254  False    False   False      False         False     False   
1255  False    False   False      False         False     False   
1256  False    False   False      False         False     False   
1257  False    False   False      False         False     False   
1258  False    False   False      False         False     False   

      family_history  mental_health_consequence  mental_health_interview  \
0              False                      False                    False   
1              False                      False                    False   
2              False                      False                    False   
3              False                      False                    False   
4              False                      False                    False   
...              ...                        ...                      ...   
1254           False                      False                    False   
1255           False                      False                    False   
1256           False                      False                    False   
1257           False                      False                    False   
1258           False                      False                    False   

      mental_vs_physical  ...  phys_health_consequence  phys_health_interview  \
0                  False  ...                    False                  False   
1                  False  ...                    False                  False   
2                  False  ...                    False                  False   
3                  False  ...                    False                  False   
4                  False  ...                    False                  False   
...                  ...  ...                      ...                    ...   
1254               False  ...                    False                  False   
1255               False  ...                    False                  False   
1256               False  ...                    False                  False   
1257               False  ...                    False                  False   
1258               False  ...                    False                  False   

      remote_work  seek_help  self_employed  state  tech_company  treatment  \
0           False      False          False  False         False      False   
1           False      False          False  False         False      False   
2           False      False          False  False         False      False   
3           False      False          False  False         False      False   
4           False      False          False  False         False      False   
...           ...        ...            ...    ...           ...        ...   
1254        False      False          False  False         False      False   
1255        False      False          False  False         False      False   
1256        False      False          False  False         False      False   
1257        False      False          False  False         False      False   
1258        False      False          False  False         False      False   

      wellness_program  work_interfere  
0                False           False  
1                False           False  
2                False           False  
3                False           False  
4                False           False  
...                ...             ...  
1254             False           False  
1255             False           False  

In [13]:
max(dataset["Age"])


99999999999

In [14]:
min(dataset["Age"])

-1726

Since cannot be 9999999999 and -1729 so we will remove theres values and will limit the vale in the age column from 0 to 100. If ther are any extreme values, then we will replace them with the median.

In [15]:
dataset["Age"].median()

31.0

In [16]:
dataset["Age"]=np.where(dataset["Age"]>100,31.0, dataset["Age"])

In [17]:
dataset["Age"]=np.where(dataset["Age"]<0,31.0, dataset["Age"])

In [18]:
dataset["Age"].describe()

count    1259.000000
mean       32.015091
std         7.360613
min         5.000000
25%        27.000000
50%        31.000000
75%        36.000000
max        72.000000
Name: Age, dtype: float64

# Now check for the Gender

In [19]:
dataset["Gender"].unique()

array(['Female', 'M', 'Male', 'male', 'female', 'm', 'Male-ish', 'maile',
       'Trans-female', 'Cis Female', 'F', 'something kinda male?',
       'Cis Male', 'Woman', 'f', 'Mal', 'Male (CIS)', 'queer/she/they',
       'non-binary', 'Femake', 'woman', 'Make', 'Nah', 'All', 'Enby',
       'fluid', 'Genderqueer', 'Female ', 'Androgyne', 'Agender',
       'cis-female/femme', 'Guy (-ish) ^_^', 'male leaning androgynous',
       'Male ', 'Man', 'Trans woman', 'msle', 'Neuter', 'Female (trans)',
       'queer', 'Female (cis)', 'Mail', 'cis male', 'A little about you',
       'Malr', 'p', 'femail', 'Cis Man',
       'ostensibly male, unsure what that really means'], dtype=object)

The column contains different names. example: Females are represented by female, Female, F, f, etc. So, set the values of the column gender to either Male or Female

In [29]:
dataset["Gender"].replace({
    "male":"Male",
    "maile":"Male",
    "m":"Male",
    "M":"Male",
    "Male-ish" :"Male",
    "something kinda male?":"Male",
    "Cis Male":"Male",
    "Mal" :"Male",
    "Male (CIS)":"Male",
    "Make":"Male",
    "Guy(-ish)" :"Male",
    "male leaning androgynous":"Male",
    "Man":"Male",
    "msle":"Male",
    "Mail" :"Male",
    "cis male":"Male",
    "Malr" :"Male",
    "Cis Man" :"Male",
    "ostensibly male":"Male"}, inplace =True)

Similiarly do it for Female

In [32]:
dataset["Gender"].replace({
    "female":"Female",
    "Trans-female" :"Female",
    "Cis Female":"Female",
    "F" "Woman" :"Female",
    "f":"Female",
    "queer/she/they":"Female",
    "Femake":"Female",
    "woman":"Female",
    "cis-female/femme":"Female",
    "Trans woman":"Female",
    "Female (trans)" :"Female",
    "Female (cis)":"Female",
    "femail":"Female"},inplace =True
    )

In [33]:
print(dataset["Gender"])

0       Female
1         Male
2         Male
3         Male
4         Male
         ...  
1254      Male
1255      Male
1256      Male
1257    Female
1258      Male
Name: Gender, Length: 1259, dtype: object


# Country

In [22]:
dataset["Country"].unique()

array(['United States', 'Canada', 'United Kingdom', 'Bulgaria', 'France',
       'Portugal', 'Netherlands', 'Switzerland', 'Poland', 'Australia',
       'Germany', 'Russia', 'Mexico', 'Brazil', 'Slovenia', 'Costa Rica',
       'Austria', 'Ireland', 'India', 'South Africa', 'Italy', 'Sweden',
       'Colombia', 'Latvia', 'Romania', 'Belgium', 'New Zealand',
       'Zimbabwe', 'Spain', 'Finland', 'Uruguay', 'Israel',
       'Bosnia and Herzegovina', 'Hungary', 'Singapore', 'Japan',
       'Nigeria', 'Croatia', 'Norway', 'Thailand', 'Denmark',
       'Bahamas, The', 'Greece', 'Moldova', 'Georgia', 'China',
       'Czech Republic', 'Philippines'], dtype=object)

Column Country has no errors

# self_employed  

In [35]:
dataset["self_employed"].unique()

array(['No', 'Yes'], dtype=object)

No need to do any changes in self_employed column

# state 

In [39]:
dataset["state"].unique()

array(['IL', 'IN', '0', 'TX', 'TN', 'MI', 'OH', 'CA', 'CT', 'MD', 'NY',
       'NC', 'MA', 'IA', 'PA', 'WA', 'WI', 'UT', 'NM', 'OR', 'FL', 'MN',
       'MO', 'AZ', 'CO', 'GA', 'DC', 'NE', 'WV', 'OK', 'KS', 'VA', 'NH',
       'KY', 'AL', 'NV', 'NJ', 'SC', 'VT', 'SD', 'ID', 'MS', 'RI', 'WY',
       'LA', 'ME'], dtype=object)

The rows whose states are known are taken as 0 by default

# family_history

In [40]:
dataset["family_history"].unique()

array(['No', 'Yes'], dtype=object)

No changes in family history

# treatment

In [54]:
dataset["treatment"].unique()

array(['Yes', 'No'], dtype=object)

No changes in treatment

# Check for other columns also

In [55]:
dataset["no_employees"].unique()

array(['Jun-25', 'More than 1000', '26-100', '100-500', '01-May',
       '500-1000'], dtype=object)

In [56]:
dataset["no_employees"].replace({
    'Jun-25': "7-25",
    '01-May':"1-5"
},inplace = True)

In [57]:
dataset["remote_work"].unique()

array(['No', 'Yes'], dtype=object)

In [58]:
dataset["tech_company"].unique()

array(['Yes', 'No'], dtype=object)

In [59]:
dataset["wellness_program"].unique()

array(['No', "Don't know", 'Yes'], dtype=object)

In [60]:
dataset["seek_help"].unique()

array(['Yes', "Don't know", 'No'], dtype=object)

In [61]:
dataset["mental_health_consequence"].unique()

array(['No', 'Maybe', 'Yes'], dtype=object)

In [63]:
dataset["phys_health_consequence"].unique()

array(['No', 'Yes', 'Maybe'], dtype=object)

In [64]:
dataset["mental_health_interview"].unique()

array(['No', 'Yes', 'Maybe'], dtype=object)

It can be seen that now the whole data is cleaned and we continue with the next steps.

# Summarize the data

In [65]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1259 entries, 0 to 1258
Data columns (total 22 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Timestamp                  1259 non-null   object 
 1   Age                        1259 non-null   float64
 2   Gender                     1259 non-null   object 
 3   Country                    1259 non-null   object 
 4   state                      1259 non-null   object 
 5   self_employed              1259 non-null   object 
 6   family_history             1259 non-null   object 
 7   treatment                  1259 non-null   object 
 8   work_interfere             1259 non-null   object 
 9   no_employees               1259 non-null   object 
 10  remote_work                1259 non-null   object 
 11  tech_company               1259 non-null   object 
 12  care_options               1259 non-null   object 
 13  wellness_program           1259 non-null   objec

In [66]:
dataset.describe()

Age
count  1259.000000
mean     32.015091
std       7.360613
min       5.000000
25%      27.000000
50%      31.000000
75%      36.000000
max      72.000000

In [68]:
dataset

Timestamp   Age  Gender         Country state self_employed  \
0     27-08-2014 11:29  37.0  Female   United States    IL            No   
1     27-08-2014 11:29  44.0    Male   United States    IN            No   
2     27-08-2014 11:29  32.0    Male          Canada     0            No   
3     27-08-2014 11:29  31.0    Male  United Kingdom     0            No   
4     27-08-2014 11:30  31.0    Male   United States    TX            No   
...                ...   ...     ...             ...   ...           ...   
1254  12-09-2015 11:17  26.0    Male  United Kingdom     0            No   
1255  26-09-2015 01:07  32.0    Male   United States    IL            No   
1256  07-11-2015 12:36  34.0    Male   United States    CA            No   
1257  30-11-2015 21:25  46.0  Female   United States    NC            No   
1258  01-02-2016 23:04  25.0    Male   United States    IL            No   

     family_history treatment work_interfere    no_employees  ...  \
0                No       Yes          Often            7-25  ...   
1                No        No         Rarely  More than 1000  ...   
2                No        No         Rarely            7-25  ...   
3               Yes       Yes          Often          26-100  ...   
4                No        No          Never         100-500  ...   
...             ...       ...            ...             ...  ...   
1254             No       Yes        No work          26-100  ...   
1255            Yes       Yes          Often          26-100  ...   
1256            Yes       Yes      Sometimes  More than 1000  ...   
1257             No        No        No work         100-500  ...   
1258            Yes       Yes      Sometimes          26-100  ...   

     care_options wellness_program   seek_help mental_health_consequence  \
0        Not sure               No         Yes                        No   
1              No       Don't know  Don't know                     Maybe   
2              No               No          No                        No   
3             Yes               No          No                       Yes   
4              No       Don't know  Don't know                        No   
...           ...              ...         ...                       ...   
1254           No               No          No                        No   
1255          Yes               No          No                        No   
1256          Yes               No          No                       Yes   
1257          Yes               No          No                       Yes   
1258          Yes               No          No                     Maybe   

     phys_health_consequence mental_health_interview phys_health_interview  \
0                         No                      No                 Maybe   
1                         No                      No                    No   
2                         No                     Yes                   Yes   
3                        Yes                   Maybe                 Maybe   
4                         No                     Yes                   Yes   
...                      ...                     ...                   ...   
1254                      No                      No                    No   
1255                      No                      No                    No   
1256                     Yes                      No                    No   
1257                      No                      No                    No   
1258                      No                      No                    No   

     mental_vs_physical obs_consequence     comments  
0                   Yes              No  No comments  
1            Don't know              No  No comments  
2                    No              No  No comments  
3                    No             Yes  No comments  
4            Don't know              No  No comments  
...                 ...             ...          ...  
1254         Don't know             

In [97]:
group_by_data=dataset.groupby(["Age"])

In [98]:
group_by_data.describe()

Timestamp                               Gender         \
         count unique               top freq  count unique   
Age                                                          
5.0          1      1  28-08-2014 10:35    1      1      1   
8.0          1      1  29-08-2014 09:10    1      1      1   
11.0         1      1  29-08-2014 17:26    1      1      1   
18.0         7      7  27-08-2014 11:43    1      7      2   
19.0         9      9  28-08-2014 09:02    1      9      2   
20.0         6      6  28-08-2014 05:05    1      6      3   
21.0        16     16  27-08-2014 18:56    1     16      4   
22.0        21     21  27-08-2014 19:13    1     21      3   
23.0        51     51  27-08-2014 13:39    1     51      4   
24.0        46     46  27-08-2014 15:50    1     46      4   
25.0        61     61  27-08-2014 12:43    1     61      5   
26.0        75     73  27-08-2014 15:23    2     75      4   
27.0        71     69  27-08-2014 12:34    2     71      3   
28.0        68     66  27-08-2014 15:32    2     68      4   
29.0        85     82  27-08-2014 15:25    2     85      4   
30.0        63     61  27-08-2014 15:31    2     63      4   
31.0        72     71  27-08-2014 15:24    2     72      7   
32.0        82     79  28-08-2014 17:02    2     82      5   
33.0        70     69  27-08-2014 15:27    2     70      4   
34.0        65     65  27-08-2014 16:14    1     65      2   
35.0        55     52  27-08-2014 11:37    2     55      2   
36.0        37     36  28-08-2014 16:52    2     37      4   
37.0        43     43  28-08-2014 12:07    1     43      4   
38.0        39     38  27-08-2014 13:24    2     39      2   
39.0        33     32  27-08-2014 12:37    2     33      3   
40.0        33     32  28-08-2014 16:57    2     33      4   
41.0        21     21  29-08-2014 13:23    1     21      2   
42.0        20     20  21-02-2015 09:30    1     20      3   
43.0        28     28  28-08-2014 16:42    1     28      2   
44.0        11     11  28-08-2014 16:09    1     11      3   
45.0        12     11  27-08-2014 12:51    2     12      3   
46.0        12     12  27-08-2014 16:36    1     12      2   
47.0         2      2  27-08-2014 16:40    1      2      2   
48.0         6      6  28-08-2014 18:17    1      6      2   
49.0         4      4  28-08-2014 09:53    1      4      1   
50.0         6      6  27-08-2014 20:52    1      6      1   
51.0         5      5  28-08-2014 17:01    1      5      1   
53.0         1      1  28-08-2014 18:21    1      1      1   
54.0         3      3  27-08-2014 16:28    1      3      1   
55.0         3      3  29-08-2014 05:47    1      3      1   
56.0         4      4  06-11-2014 11:24    1      4      3   
57.0         3      3  27-08-2014 16:06    1      3      1   
58.0         1      1  27-08-2014 16:13    1      1      1   
60.0         2      2  24-02-2015 10:32    1      2      1   
61.0         1      1  29-08-2014 01:20    1      1      1   
62.0         1      1  27-08-2014 17:12    1      1      1   
65.0         1      1  27-08-2014 19:17    1      1      1   
72.0         1      1  02-10-2014 21:25    1      1      1   

                              Country         ... mental_vs_physical       \
                     top freq   count unique  ...                top freq   
Age                                           ...                           
5.0                 Male    1       1      1  ...                Yes    1   
8.0   A little about you    1       1      1  ...                Yes    1   
11.0                Male    1       1      1  ...                Yes    1   
18.0                Male    6       7      5  ...         Don't know    5   
19.0                Male    8       9      4  ...         Don't know    5   
20.0                Male    4       6      5  ...                Yes    3   
21.0                Male   13      16      4  ...         Don't know    8   
22.0                Male   15      21      8  ...         Don't know    